### ISSUE TITLE
[MODULES-372] JDK 11 - module not found java.se

### ISSUE THREAD

**Issue by dmlloyd** [jboss-modules/jboss-modules#543]
@manovotn opened **[MODULES-372](https://issues.redhat.com/browse/MODULES-372?redirect=false)** and commented

When trying to start up WFLY with JDK 11 (using `standalone.sh`, I am getting this error:

```
org.jboss.modules.ModuleNotFoundException: java.se
	at org.jboss.modules.Module.addPaths(Module.java:1266)
	at org.jboss.modules.Module.link(Module.java:1622)
	at org.jboss.modules.Module.relinkIfNecessary(Module.java:1650)
	at org.jboss.modules.ModuleLoader.loadModule(ModuleLoader.java:296)
	at org.jboss.modules.Main.main(Main.java:437)
```

Output of `java -version`:

```
openjdk version "11-ea" 2018-09-25
OpenJDK Runtime Environment 18.9 (build 11-ea+21)
OpenJDK 64-Bit Server VM 18.9 (build 11-ea+21, mixed mode)
```

---

**Affects:** 1.8.5.Final

**Issue Links:**
- [JBWS-4129](https://issues.redhat.com/browse/JBWS-4129) Can't deploy webservice on JDK11
 (_**"blocks"**_)
- [ENTESB-10072](https://issues.redhat.com/browse/ENTESB-10072) Can't install fuse on EAP 7.2 with Java 11
 (_**"blocks"**_)
- [WELD-2516](https://issues.redhat.com/browse/WELD-2516) Weld TS execution with JDK 11
 (_**"blocks"**_)
- [WFCORE-4211](https://issues.redhat.com/browse/WFCORE-4211) Missing docs: Embedded container does not work without the JVM being launched with java.se module on Java 11
 (_**"blocks"**_)
- [WINDUP-2149](https://issues.redhat.com/browse/WINDUP-2149) Run RHAMT on JDK 11
 (_**"is blocked by"**_)
- [WFCORE-3962](https://issues.redhat.com/browse/WFCORE-3962) Starting WFLY scripts with JDK 11 blows up with ModuleNotFoundException java.se

-1 votes, 17 watchers

**Comment by dmlloyd** [comment]
@dmlloyd commented

This change to the JDK: http://cr.openjdk.java.net/~alanb/8197532/webrev/ which I supported strongly ( :( ) causes java.se to not be included in the set of modules in class path mode by default.  The JDK has to be launched with `--add-module java.se` to work around the problem.

**Comment by dmlloyd** [comment]
**[Kabir Khan](https://issues.redhat.com/secure/ViewProfile.jspa?name=kabirkhan)** commented

[dlloyd@redhat.com](https://issues.redhat.com/secure/ViewProfile.jspa?name=dlloyd@redhat.com) Should this be closed? As I believe we have worked around this as recommended in issues like:
* https://issues.jboss.org/browse/WFLY-10937
* https://issues.jboss.org/browse/WFLY-10811
* https://issues.jboss.org/browse/WFLY-10772

**Comment by dmlloyd** [comment]
@dmlloyd commented

I'd like to figure out a "real" fix at some point, so let's leave this open for now.

**Comment by dmlloyd** [comment]
**[Francesco Marchioni](https://issues.redhat.com/secure/ViewProfile.jspa?name=f_marchioni)** commented

Thanks for the workaround (I've verified it on OpenJDK build 11.0.1). Just a thing, the correct parameter to add is **--add-modules java.se**

**Comment by dmlloyd** [comment]
**[Wolfgang Mayer](https://issues.redhat.com/secure/ViewProfile.jspa?name=mayerw01)** commented

This work around does not work on Java HotSpot

**Comment by dmlloyd** [comment]
@dmlloyd commented

Could you please elaborate?

**Comment by dmlloyd** [comment]
**[Wolfgang Mayer](https://issues.redhat.com/secure/ViewProfile.jspa?name=mayerw01)** commented

Sorry I accidentally misplaced the parameter. I've now changed the appclient.sh directly to

eval \"$JAVA\" $JAVA_OPTS \
--add-modules java.se \
-cp "$CLASSPATH" \ ...

This seems to work

**Comment by dmlloyd** [comment]
**[George Gastaldi](https://issues.redhat.com/secure/ViewProfile.jspa?name=gastaldi)** commented

[dlloyd@redhat.com](https://issues.redhat.com/secure/ViewProfile.jspa?name=dlloyd@redhat.com) how about falling back to a JBoss Module if the `java.se` JDK module cannot be found? (With a flag to enable this behavior)

**Comment by dmlloyd** [comment]
@Ladicek commented

My understanding is that it's JBoss Modules who implicitly add the dependency on `java.se` to \[almost\] all modules. While the `java.se` module isn't "available" by default, this module is nothing but an aggregator of other modules that _are_ "available" by default (I think -- please correct me if I'm wrong here). So technically, JBoss Modules could either add an implicit dependency on all these modules, or create another aggregating module and add a dependency on that. (Do I even make sense here?)

**Comment by dmlloyd** [comment]
@dmlloyd commented

Correct. We could re-aggregate the relevant modules by hand.

**Comment by dmlloyd** [comment]
@Ladicek commented

I took a stab at this in https://github.com/Ladicek/jboss-modules/commits/emulated-java-se-module and would appreciate any comments. Especially if this approach has any chance of being accepted :-)

The one question I wasn't able to answer on my own is: should the `org.jboss.modules.emulated.java.se` module have its own distinct name, or should it be called `java.se`. I opted into the former option, as it seems less invasive. The latter option would be more transparent, but might have unintended consequences that I'm not aware of.

With the current name, I was able to confirm with Thorntail that adding `--add-modules java.se` is no longer necessary, but only after I changed the definition of the `javax.api` module from WildFly (as that refers directly to the `java.se` name).
